File: src/main/java/com/arnav/tutionSAAS/service/AdminService.java
package com.arnav.tutionSAAS.service;

import com.arnav.tutionSAAS.dto.FeeStatusResponse;
import com.arnav.tutionSAAS.dto.InviteRequest;
import com.arnav.tutionSAAS.dto.InviteResponse;
import com.arnav.tutionSAAS.entity.*;
import com.arnav.tutionSAAS.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class AdminService {

    @Autowired private UserRepo userRepo;
    @Autowired private StudentRepo studentRepo;
    @Autowired private TeacherInviteRepo inviteRepo;
    @Autowired private BatchRepo batchRepo;

    @Value("${app.frontend.url:http://localhost:3000}")
    private String frontendUrl;

    // ─── Teacher Invite ───

    @Transactional
    public InviteResponse generateTeacherInvite(InviteRequest request) {
        TeacherInvite invite = new TeacherInvite();
        invite.setToken(UUID.randomUUID().toString());
        invite.setEmail(request != null ? request.getEmail() : null);
        invite.setUsed(false);
        invite.setCreatedAt(LocalDateTime.now());
        invite.setExpiresAt(LocalDateTime.now().plusDays(7));

        TeacherInvite saved = inviteRepo.save(invite);
        return toInviteResponse(saved);
    }

    public List<InviteResponse> getAllInvites() {
        return inviteRepo.findAll().stream()
                .map(this::toInviteResponse)
                .collect(Collectors.toList());
    }

    // ─── Fee Management ───

    public List<FeeStatusResponse> getAllFeeStatus() {
        return userRepo.findByRole(Role.STUDENT).stream()
                .map(student -> {
                    FeeStatusResponse dto = new FeeStatusResponse();
                    dto.setStudentId(student.getId());
                    dto.setStudentName(student.getFullName());
                    dto.setEmail(student.getEmail());
                    dto.setGrade(student.getGrade());
                    dto.setBlocked(student.isBlocked());

                    StudentProfile profile = studentRepo.findById(student.getId()).orElse(null);
                    dto.setFeesPaid(profile != null && profile.isFeesPaidForCurrentMonth());

                    List<String> batchNames = batchRepo.findByStudents_Id(student.getId())
                            .stream().map(Batch::getName).collect(Collectors.toList());
                    dto.setBatchNames(batchNames);

                    return dto;
                })
                .collect(Collectors.toList());
    }

    @Transactional
    public void blockStudent(Long studentId) {
        User user = userRepo.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));
        if (user.getRole() != Role.STUDENT) {
            throw new RuntimeException("Can only block student accounts");
        }
        user.setBlocked(true);
        userRepo.save(user);
    }

    @Transactional
    public void unblockStudent(Long studentId) {
        User user = userRepo.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));
        user.setBlocked(false);
        userRepo.save(user);
    }

    @Transactional
    public void setMonthlyFee(Long batchId, double amount) {
        Batch batch = batchRepo.findById(batchId)
                .orElseThrow(() -> new RuntimeException("Batch not found"));
        if (amount < 0) throw new RuntimeException("Fee amount cannot be negative");
        batch.setMonthlyFee(amount);
        batchRepo.save(batch);
    }

    // ─── Dashboard ───

    public DashboardStats getDashboardStats() {
        DashboardStats stats = new DashboardStats();
        stats.setTotalStudents(userRepo.findByRole(Role.STUDENT).size());
        stats.setTotalTeachers(userRepo.findByRole(Role.TEACHER).size());
        stats.setTotalParents(userRepo.findByRole(Role.PARENT).size());
        stats.setActiveBatches(batchRepo.findByIsActiveTrue().size());

        long unpaid = userRepo.findByRole(Role.STUDENT).stream()
                .filter(s -> {
                    StudentProfile p = studentRepo.findById(s.getId()).orElse(null);
                    return p == null || !p.isFeesPaidForCurrentMonth();
                }).count();
        stats.setUnpaidFeesCount(unpaid);

        long blocked = userRepo.findByRole(Role.STUDENT).stream()
                .filter(User::isBlocked).count();
        stats.setBlockedStudentsCount(blocked);

        return stats;
    }

    // ─── Helpers ───

    private InviteResponse toInviteResponse(TeacherInvite invite) {
        InviteResponse dto = new InviteResponse();
        dto.setId(invite.getId());
        dto.setToken(invite.getToken());
        dto.setInviteUrl(frontendUrl + "/signup?invite=" + invite.getToken());
        dto.setEmail(invite.getEmail());
        dto.setUsed(invite.isUsed());
        dto.setExpiresAt(invite.getExpiresAt());
        dto.setCreatedAt(invite.getCreatedAt());
        return dto;
    }

    // ─── Inner class for dashboard ───

    @lombok.Data
    public static class DashboardStats {
        private long totalStudents;
        private long totalTeachers;
        private long totalParents;
        private long activeBatches;
        private long unpaidFeesCount;
        private long blockedStudentsCount;
    }
}

File: backend-nest/src/admin/admin.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { AuditService } from '../audit/audit.service';
import { Role } from '@prisma/client';
import { v4 as uuidv4 } from 'uuid';

@Injectable()
export class AdminService {
  constructor(
    private prisma: PrismaService,
    private config: ConfigService,
    private audit: AuditService,
  ) {}

  async generateTeacherInvite(teacherName?: string) {
    const token = uuidv4();
    const expiresAt = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);

    const invite = await this.prisma.teacherInvite.create({
      data: { token, used: false, expiresAt, teacherName: teacherName || null },
    });

    await this.audit.log('INVITE_CREATED', undefined, 'TeacherInvite', `id=${invite.id}`);
    return toInviteResponse(invite, this.config.get('FRONTEND_URL'));
  }

  async deleteInvite(id: number) {
    const invite = await this.prisma.teacherInvite.findUnique({ where: { id } });
    if (!invite) throw new NotFoundException('Invite not found');
    await this.prisma.teacherInvite.delete({ where: { id } });
    await this.audit.log('INVITE_DELETED', undefined, 'TeacherInvite', `id=${id}`);
  }

  async getAllInvites() {
    const invites = await this.prisma.teacherInvite.findMany({
      orderBy: { createdAt: 'desc' },
    });
    const frontendUrl = this.config.get('FRONTEND_URL');
    return invites.map((i) => toInviteResponse(i, frontendUrl));
  }

  async getAllFeeStatus() {
    const students = await this.prisma.user.findMany({
      where: { role: Role.STUDENT },
      include: {
        studentProfile: true,
        enrolledBatches: true,
        linkedByParents: { include: { user: true } },
      },
    });

    return students.map((student) => ({
      studentId: student.id,
      studentName: student.fullName,
      email: student.email,
      grade: student.grade,
      studentPhone: student.phoneNumber,
      blocked: student.blocked,
      feesPaid: student.studentProfile?.feesPaidForCurrentMonth ?? false,
      batchNames: student.enrolledBatches.map((b) => b.name),
      parentPhone: student.linkedByParents[0]?.user?.phoneNumber || null,
      parentEmail: student.linkedByParents[0]?.user?.email || null,
    }));
  }

  async blockStudent(studentId: number) {
    const user = await this.prisma.user.findUnique({ where: { id: studentId } });
    if (!user) throw new NotFoundException('Student not found');
    if (user.role !== Role.STUDENT) throw new NotFoundException('Can only block student accounts');

    await this.prisma.user.update({
      where: { id: studentId },
      data: { blocked: true },
    });
    await this.audit.log('STUDENT_BLOCKED', undefined, 'User', `studentId=${studentId}`);
  }

  async unblockStudent(studentId: number) {
    const user = await this.prisma.user.findUnique({ where: { id: studentId } });
    if (!user) throw new NotFoundException('Student not found');

    await this.prisma.user.update({
      where: { id: studentId },
      data: { blocked: false },
    });
    await this.audit.log('STUDENT_UNBLOCKED', undefined, 'User', `studentId=${studentId}`);
  }

  async setMonthlyFee(batchId: number, amount: number) {
    if (amount < 0) throw new Error('Fee amount cannot be negative');

    await this.prisma.batch.update({
      where: { id: batchId },
      data: { monthlyFee: amount },
    });
  }

  async getDashboardStats() {
    const [students, teachers, parents, activeBatches] = await Promise.all([
      this.prisma.user.count({ where: { role: Role.STUDENT } }),
      this.prisma.user.count({ where: { role: Role.TEACHER } }),
      this.prisma.user.count({ where: { role: Role.PARENT } }),
      this.prisma.batch.count({ where: { isActive: true } }),
    ]);

    const allStudents = await this.prisma.user.findMany({
      where: { role: Role.STUDENT },
      include: {
        studentProfile: true,
        enrolledBatches: true,
      },
    });

    const unpaidFeesCount = allStudents.filter((s) => {
      if (s.enrolledBatches.length === 0) return false;
      return !s.studentProfile?.feesPaidForCurrentMonth;
    }).length;

    const blockedStudentsCount = allStudents.filter((s) => s.blocked).length;

    return {
      totalStudents: students,
      totalTeachers: teachers,
      totalParents: parents,
      activeBatches,
      unpaidFeesCount,
      blockedStudentsCount,
    };
  }
}

function toInviteResponse(invite: any, frontendUrl: string) {
  return {
    id: invite.id,
    token: invite.token,
    teacherName: invite.teacherName ?? null,
    inviteUrl: `${frontendUrl}/sign-up?invite=${invite.token}`,
    used: invite.used,
    expiresAt: invite.expiresAt,
    createdAt: invite.createdAt,
  };
}
